subscriber.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package pubsub
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/deque"
+)
+
+var logger = loggo.GetLogger("pubsub.subscriber")
+
+type subscriber struct {
+	id int
+
+	topicMatcher TopicMatcher
+	handler      func(topic Topic, data interface{})
+
+	mutex   sync.Mutex
+	pending *deque.Deque
+	closed  chan struct{}
+	data    chan struct{}
+	done    chan struct{}
+}
+
+func newSubscriber(matcher TopicMatcher, handler interface{}) (*subscriber, error) {
+	f, err := checkHandler(handler)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Tracef("new subscriber, handler func %v", f)
+	// A closed channel is used to provide an immediate route through a select
+	// call in the loop function.
+	closed := make(chan struct{})
+	close(closed)
+	sub := &subscriber{
+		topicMatcher: matcher,
+		handler:      f,
+		pending:      deque.New(),
+		data:         make(chan struct{}, 1),
+		done:         make(chan struct{}),
+		closed:       closed,
+	}
+	go sub.loop()
+	logger.Debugf("created subscriber %p for %v", sub, matcher)
+	return sub, nil
+}
+
+func (s *subscriber) close() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	// need to iterate through all the pending calls and make sure the wait group
+	// is decremented. this isn't exposed yet, but needs to be.
+	for call, ok := s.pending.PopFront(); ok; call, ok = s.pending.PopFront() {
+		call.(*handlerCallback).done()
+	}
+	close(s.done)
+}
+
+func (s *subscriber) loop() {
+	var next <-chan struct{}
+	for {
+		select {
+		case <-s.done:
+			return
+		case <-s.data:
+			// Has new data been pushed on?
+		case <-next:
```

Comment:
@howbazaar huh? there is one select in my latest suggestion. the first example had two, because I didn't notice that you drain the queue on close. anyway, take it or leave it - just my opinion.